# HG changeset 9680 patch
From: kaf24@firebug.cl.cam.ac.uk
# Node ID 9c1f1e609d4fd82716849c50cd344f7cb7906a3e
# Parent  7ae0f83cf355525e5c14832e77df9b13f67ee92c
Subject: deal with missing media in CD drive exported to guest

Prevent an oops in Dom0 that occurs when a CD device, specified as one
of the 'hardrives' in the 'disk=' line of a para-virtualized guest's
def file, has no media when the guest is started.

The oops occurs in vbd.c when vbd_size() is called from connect() (in
xenbus.c) and the vbd pointer is really an error code that comes from
the failed open that occurred in vbd_create().

Based on a patch from Ross Maxfield at Novell.

Signed-off-by: Keir Fraser <keir@xensource.com>
xen-unstable changeset:   9998:df1521633fb519727adb80f15ea1f2e5b69dc3df
xen-unstable date:        Tue May 16 09:40:38 2006 +0100

Acked-By: Jan Beulich <jbeulich@novell.com>
References: 175114

Index: head-2006-05-19/drivers/xen/blkback/vbd.c
===================================================================
--- head-2006-05-19.orig/drivers/xen/blkback/vbd.c	2006-05-19 10:36:16.000000000 +0200
+++ head-2006-05-19/drivers/xen/blkback/vbd.c	2006-05-19 10:36:56.000000000 +0200
@@ -55,6 +55,7 @@ int vbd_create(blkif_t *blkif, blkif_vde
 	       unsigned minor, int readonly)
 {
 	struct vbd *vbd;
+	struct block_device *bdev;
 
 	vbd = &blkif->vbd;
 	vbd->handle   = handle; 
@@ -63,15 +64,17 @@ int vbd_create(blkif_t *blkif, blkif_vde
 
 	vbd->pdevice  = MKDEV(major, minor);
 
-	vbd->bdev = open_by_devnum(
-		vbd->pdevice,
-		vbd->readonly ? FMODE_READ : FMODE_WRITE);
-	if (IS_ERR(vbd->bdev)) {
-		DPRINTK("vbd_creat: device %08x doesn't exist.\n",
+	bdev = open_by_devnum(vbd->pdevice,
+			      vbd->readonly ? FMODE_READ : FMODE_WRITE);
+
+	if (IS_ERR(bdev)) {
+		DPRINTK("vbd_creat: device %08x could not be opened.\n",
 			vbd->pdevice);
 		return -ENOENT;
 	}
 
+	vbd->bdev = bdev;
+
 	if (vbd->bdev->bd_disk == NULL) {
 		DPRINTK("vbd_creat: device %08x doesn't exist.\n",
 			vbd->pdevice);
